flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManager.java

```diff
@@ -0,0 +1,697 @@
+            sourceNodes.add(streamGraph.getStreamNode(sourceNodeId));
+        }
+        if (jobGraph.isDynamic()) {
+            setVertexParallelismsForDynamicGraphIfNecessary(sourceNodes);
+        }
+        createJobVerticesAndUpdateGraph(sourceNodes);
+    }
+
+    private List<JobVertex> createJobVerticesAndUpdateGraph(List<StreamNode> streamNodes) {
+        final JobVertexBuildContext jobVertexBuildContext =
+                new JobVertexBuildContext(
+                        jobGraph,
+                        streamGraph,
+                        hasHybridResultPartition,
+                        hashes,
+                        legacyHashes,
+                        defaultSlotSharingGroup);
+
+        createOperatorChainInfos(streamNodes, jobVertexBuildContext);
+
+        recordCreatedJobVerticesInfo(jobVertexBuildContext);
+
+        generateConfigForJobVertices(jobVertexBuildContext);
+
+        return new ArrayList<>(jobVertexBuildContext.getJobVerticesInOrder().values());
+    }
+
+    private void generateConfigForJobVertices(JobVertexBuildContext jobVertexBuildContext) {
+        // this may be used by uncreated down stream vertex.
+        final Map<Integer, Map<StreamEdge, NonChainedOutput>> opIntermediateOutputs =
+                new HashMap<>();
+
+        setAllOperatorNonChainedOutputsConfigs(opIntermediateOutputs, jobVertexBuildContext);
+
+        setAllVertexNonChainedOutputsConfigs(opIntermediateOutputs, jobVertexBuildContext);
+
+        connectToFinishedUpStreamVertex(jobVertexBuildContext);
+
+        setPhysicalEdges(jobVertexBuildContext);
+
+        markSupportingConcurrentExecutionAttempts(jobVertexBuildContext);
+
+        validateHybridShuffleExecuteInBatchMode(jobVertexBuildContext);
+
+        // When generating in a single step, there may be differences between the results and the
+        // full image generation. We consider this difference to be normal because they do not need
+        // to be in the same shared group.
+        setSlotSharingAndCoLocation(jobVertexBuildContext);
+
+        setManagedMemoryFraction(jobVertexBuildContext);
+
+        addVertexIndexPrefixInVertexName(jobVertexBuildContext, vertexIndexId);
+
+        setVertexDescription(jobVertexBuildContext);
+
+        serializeOperatorCoordinatorsAndStreamConfig(serializationExecutor, jobVertexBuildContext);
+    }
+
+    private void setAllVertexNonChainedOutputsConfigs(
+            final Map<Integer, Map<StreamEdge, NonChainedOutput>> opIntermediateOutputs,
+            JobVertexBuildContext jobVertexBuildContext) {
+        jobVertexBuildContext
+                .getJobVerticesInOrder()
+                .keySet()
+                .forEach(
+                        startNodeId ->
+                                setVertexNonChainedOutputsConfig(
+                                        startNodeId, opIntermediateOutputs, jobVertexBuildContext));
+    }
+
+    private void setVertexNonChainedOutputsConfig(
```

Comment:
It's better to explain in method level comments about the difference from that in `StreamJobGraphGenerator`.